static/src/javascripts-legacy/projects/common/modules/onward/tech-feedback.js

```diff
@@ -110,23 +51,89 @@ define([
+            } else if(elem.value !== "nothing") {
+                document.getElementById(elem.value).classList.remove("feedback__form--selected");
+            }
+        });
+    }
+
+    function mandatoryCheck(elem) {
+        if(elem.value === ""){
+            elem.classList.add("feedback__entry--mandatory-failed");
+            return false;
+        } else {
+            elem.classList.remove("feedback__entry--mandatory-failed");
+            return true;
+        }
+    }
+
+    function initForms() {
+
+        var warning = document.getElementById("feedback__explainer");
+
+        // mandatory checks (realtime)
+
+        $.forEachElement(".feedback__form input,.feedback__form textarea", function(elem){
+            elem.addEventListener("blur", function(){ mandatoryCheck(elem); });
+            elem.addEventListener("input", function(){ mandatoryCheck(elem); });
+        });
+
+        // mandatory checks (on submit)
+
+        $.forEachElement(".feedback__form form", function(elem){
+            elem.addEventListener("submit", function() {
+
+                var hasFailed = false;
+
+                $.forEachElement(".feedback__form--selected input,.feedback__form--selected textarea", function(elem){
+                    if(!mandatoryCheck(elem)){
+                        hasFailed = true;
+                    }
+                });
+
+                if(hasFailed){
+                    warning.innerHTML = "All fields must be filled to proceed";
+                }
+
+                return !hasFailed;
+
+            });
+        });
+
+        // form toggling
+
+        document.getElementById("feedback-category").addEventListener("change", toggleFormVisibility, false);
+
+        // insert hidden extra data into forms
+
+        $.forEachElement(".feedback__form input[name=extra]", function(elem){
+            elem.value = JSON.stringify(getExtraDataInformation());
+        })
+
+    }
+
+    function hideUnenhancedFallback() {
+        document.getElementById("feedback-form-default").remove();
+    }
+
     return function () {
+
         detect.adblockInUse.then(function(adblockInUse){
             adblockBeingUsed = adblockInUse;
         });
 
-        var storedValues = getValuesFromHash(window.location.hash);
-        registerHandler('.js-tech-feedback-report', addEmailValuesToHash(storedValues));
-        registerHandler('.js-tech-feedback-mailto', addEmailHeaders(storedValues));
-        registerHandler('[href=mailto:userhelp@theguardian.com]', addEmailHeaders(storedValues));
-        registerHandler('[href=mailto:crosswords.beta@theguardian.com]', addEmailHeaders(storedValues));// FIXME should have used a .js- selector
+        initForms();
+        hideUnenhancedFallback();
 
-        // Exposed for testing
-        this.getValuesFromHash = getValuesFromHash;
+        // exposed for testing
+
+        this.getExtraDataInformation = getExtraDataInformation;
         this.summariseAbTests = summariseAbTests;
```

Comment:
Rather than expose these functions for testing, I think it would be nicer to test whether the expected data has been stringified and set as the value on `.feedback__form input[name=extra]` (i.e. that lines 113-115 have executed correctly)